[JSC] Use JumpTable in WasmBBQJIT

Constellation opened the pull request:
#### 7202c59d1d29bec2ad6994e6c59aa4e27bf60f85
<pre>
[JSC] Use JumpTable in WasmBBQJIT
<a href="https://bugs.webkit.org/show_bug.cgi?id=253472">https://bugs.webkit.org/show_bug.cgi?id=253472</a>
rdar://106330476

Reviewed by Justin Michaud.

This patch implements JumpTable in WasmBBQJIT to implement wasm BrTable efficiently.
Currently, we are always doing binary-switch, which significantly bloats the code size.
Fortunately, wasm BrTable is super simple: targets are 0-N, and taking default if it exceeds N.
This is pretty easy to generate JumpTable for that.

We generate JumpTable in the link task, and put it in BBQCallee. It improves JetStream2/tsf-wasm&apos;s Runtime by 7~10%.

Before:
    Running tsf-wasm:
        Startup: 833.333
        Run time: 3.415
        Score: 53.349

After:
    Running tsf-wasm:
        Startup: 833.333
        Run time: 3.652
        Score: 55.169

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::BBQJIT):
(JSC::Wasm::BBQJIT::addSwitch):
(JSC::Wasm::parseAndCompileBBQ):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmBBQPlan.h:
* Source/JavaScriptCore/wasm/WasmCallee.h:

Canonical link: <a href="https://commits.webkit.org/261345@main">https://commits.webkit.org/261345@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12e9b32481ae43dc7f2d70350d4b392bc1b56101

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111239 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20378 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43782 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2665 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120049 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21746 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11476 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/2259 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117001 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16147 "Found 1 new test failure: fast/scrolling/keyboard-scrolling-home.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99322 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103828 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98104 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30984 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44696 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/99780 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12883 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/32319 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/86565 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/10995 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13405 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9313 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/100961 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18838 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51909 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31468 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15358 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/108998 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4308 "Built successfully and passed tests") | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26869 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (approved):
r=me

In my experience, 4 was a better heuristic than 7. Let's see once it is easier to run a perf task.

Constellation committed:
[JSC] Use JumpTable in WasmBBQJIT
https://bugs.webkit.org/show_bug.cgi?id=253472
rdar://106330476

Reviewed by Justin Michaud.

This patch implements JumpTable in WasmBBQJIT to implement wasm BrTable efficiently.
Currently, we are always doing binary-switch, which significantly bloats the code size.
Fortunately, wasm BrTable is super simple: targets are 0-N, and taking default if it exceeds N.
This is pretty easy to generate JumpTable for that.

We generate JumpTable in the link task, and put it in BBQCallee. It improves JetStream2/tsf-wasm's Runtime by 7~10%.

Before:
    Running tsf-wasm:
        Startup: 833.333
        Run time: 3.415
        Score: 53.349

After:
    Running tsf-wasm:
        Startup: 833.333
        Run time: 3.652
        Score: 55.169

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::BBQJIT):
(JSC::Wasm::BBQJIT::addSwitch):
(JSC::Wasm::parseAndCompileBBQ):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmBBQPlan.h:
* Source/JavaScriptCore/wasm/WasmCallee.h:

Canonical link: https://commits.webkit.org/261345@main
